vault/namespaces_openbao.go

```diff
@@ -0,0 +1,475 @@
+		if err != nil {
+			return handleError(err)
+		}
+
+		if ns == nil {
+			return nil, nil
+		}
+
+		resp := &logical.Response{
+			Data: map[string]interface{}{
+				"id":              ns.ID,
+				"path":            ns.Path,
+				"custom_metadata": ns.CustomMetadata,
+			},
+		}
+
+		return resp, nil
+	}
+}
+
+// handleNamespaceSet handles the "/sys/namespaces/<path>" endpoint to set a namespace
+func (b *SystemBackend) handleNamespacesSet() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+		path := data.Get("path").(string)
+		imetadata, ok := data.GetOk("custom_metadata")
+		var metadata map[string]string
+		if ok {
+			metadata = make(map[string]string)
+			for k, v := range imetadata.(map[string]interface{}) {
+				if metadata[k], ok = v.(string); !ok {
+					return nil, fmt.Errorf("custom_metadata values must be strings")
+				}
+			}
+		}
+
+		// Update the namespace
+		if err := b.Core.namespaceStore.SetNamespace(ctx, path, metadata); err != nil {
+			return handleError(err)
+		}
+
+		return nil, nil
+	}
+}
+
+// handleNamespacesPatch handles the "/sys/namespace/<path>" endpoints to update a namespace
+func (b *SystemBackend) handleNamespacesPatch() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+		path := data.Get("path").(string)
+		imetadata, ok := data.GetOk("custom_metadata")
+		var metadata map[string]string
+		if ok {
+			metadata = imetadata.(map[string]string)
+		}
+
+		// Update the namespace
+		if err := b.Core.namespaceStore.PatchNamespace(ctx, path, metadata); err != nil {
+			return handleError(err)
+		}
+
+		return nil, nil
+	}
+}
+
+func (b *SystemBackend) handleNamespacesDelete() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+		path := data.Get("path").(string)
+
+		if err := b.Core.namespaceStore.DeleteNamespace(ctx, path); err != nil {
+			return handleError(err)
+		}
+		return nil, nil
+	}
+}
+
+// NamespaceStore is used to provide durable storage of namespace
+type NamespaceStore struct {
+	core    *Core
+	aclView BarrierView
+
+	// This is used to ensure that writes to the store (acl) or to the egp
+	// path tree don't happen concurrently. We are okay reading stale data so
+	// long as there aren't concurrent writes.
+	modifyLock *sync.RWMutex
+
+	// logger is the server logger copied over from core
+	logger hclog.Logger
+}
+
+// NamespaceEntry is used to store a namespace by name
+type NamespaceEntry struct {
+	namespace.Namespace
+}
+
+// NewNamespaceStore creates a new NamespaceStore that is backed
+// using a given view. It used used to durable store and manage named namespace.
+func NewNamespaceStore(ctx context.Context, core *Core, baseView BarrierView, system logical.SystemView, logger hclog.Logger) (*NamespaceStore, error) {
```

Comment:
```suggestion
func NewNamespaceStore(ctx context.Context, core *Core, baseView BarrierView, logger hclog.Logger) (*NamespaceStore, error) {
```

`SystemView` is unused and I don't think it has a purpose in this code.

Reply:
This has ramifications elsewhere.